File: frontend/src/lib/print-job.ts
export type PrintJobStatus =
  | "SUBMITTED"
  | "REJECTED"
  | "APPROVED"
  | "QUEUED"
  | "READY_TO_PRINT"
  | "SLICING"
  | "SLICED"
  | "PRINTING"
  | "PAUSED"
  | "FINISHED"
  | "FAILED"
  | "CANCELLED"
  | "EXPIRED"
  | "DELETED";

export interface UploadedFileOut {
  id: string;
  original_filename: string;
  file_size_bytes: number;
  sha256: string;
  created_at: string;
}

export interface PrintJobOwnerOut {
  id: string;
  username: string;
  display_name: string;
  class_name: string | null;
}

export interface PrintJob {
  id: string;
  title: string;
  status: PrintJobStatus;
  queue_position: number | null;
  student_note: string | null;
  teacher_note: string | null;
  estimated_print_time_seconds: number | null;
  estimated_filament_grams: number | null;
  created_at: string;
  approved_at: string | null;
  rejected_at: string | null;
  queued_at: string | null;
  started_at: string | null;
  finished_at: string | null;
  failed_at: string | null;
  expires_at: string | null;
  uploaded_file: UploadedFileOut;
  owner?: PrintJobOwnerOut | null;
}

export const statusLabels: Record<PrintJobStatus, string> = {
  SUBMITTED: "Submitted",
  REJECTED: "Rejected",
  APPROVED: "Approved",
  QUEUED: "In queue",
  READY_TO_PRINT: "Ready to print",
  SLICING: "Slicing",
  SLICED: "Sliced",
  PRINTING: "Printing",
  PAUSED: "Paused",
  FINISHED: "Finished",
  FAILED: "Failed",
  CANCELLED: "Cancelled",
  EXPIRED: "Expired",
  DELETED: "Deleted",
};

export const statusBadgeVariant: Record<
  PrintJobStatus,
  "default" | "secondary" | "destructive" | "success" | "warning" | "outline"
> = {
  SUBMITTED: "secondary",
  REJECTED: "destructive",
  APPROVED: "default",
  QUEUED: "default",
  READY_TO_PRINT: "default",
  SLICING: "warning",
  SLICED: "default",
  PRINTING: "warning",
  PAUSED: "warning",
  FINISHED: "success",
  FAILED: "destructive",
  CANCELLED: "outline",
  EXPIRED: "outline",
  DELETED: "outline",
};

File: backend/app/schemas/print_job.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import PrintJobStatus


class UploadedFileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    original_filename: str
    file_size_bytes: int
    sha256: str
    created_at: datetime


class PrintJobOwnerOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    username: str
    display_name: str
    class_name: str | None = None


class JobActionRequest(BaseModel):
    teacher_note: str | None = None


class StartPrintRequest(BaseModel):
    printer_id: uuid.UUID


class PrintJobPrinterOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str


class PrintJobOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    status: PrintJobStatus
    queue_position: int | None = None
    student_note: str | None = None
    teacher_note: str | None = None
    estimated_print_time_seconds: int | None = None
    estimated_filament_grams: float | None = None
    created_at: datetime
    approved_at: datetime | None = None
    rejected_at: datetime | None = None
    queued_at: datetime | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    failed_at: datetime | None = None
    expires_at: datetime | None = None
    uploaded_file: UploadedFileOut
    owner: PrintJobOwnerOut | None = None
    printer: PrintJobPrinterOut | None = None


class PrintJobLiveStatusOut(BaseModel):
    status: str
    progress: float = 0.0
    nozzle_actual: float | None = None
    nozzle_target: float | None = None
    bed_actual: float | None = None
    bed_target: float | None = None
